Feature/create plans

dennisylyung opened the pull request:
Resolves #4

Elynor-Kamil commented on models/plan.py:
```diff
@@ -0,0 +1,36 @@
+        self.description = description
+        self.geographical_area = geographical_area
+        self.start_date = start_date
+        self.camps: set[Camp] = set(camps) if camps else set()
```
Should be mandatory to add at least one camp at initialisation

Elynor-Kamil replied:
If a camp is not included, the following exception can be raised:

```
    def missing_camps_error(Exception):
        """
        It is mandatory to supply at least one camp when you are creating an emergency plan.
        When no camps are provided, this exception will be raised.
        """

        def __init__(self, camps=set(), message="It is mandatory to provide at least one camp"):
            self.camps = camps
            self.message = message
            super().__init__(self.message)
```

dennisylyung replied:
@Elynor-Kamil I think you mean a class.
I'd suggest removing the camps attribute, and make the message fixed. 
i.e. 
```
    class MissingCampsError(Exception):
        """
        It is mandatory to supply at least one camp when you are creating an emergency plan.
        When no camps are provided, this exception will be raised.
        """

        def __init__(self):
            super().__init__("It is mandatory to provide at least one camp")
```
Is that okay? If so I will add it to the code

Shalaka-Yerawadekar reviewed (commented):
Ensuring that start_date is valid by including the following in class Plan:

```
       def check_date(self, start_date):
        startDate = datetime.datetime.strptime(start_date, "%d/%m/%Y")
        present = datetime.datetime.now()
        if startDate < present:
            raise InValidStartDate(start_date)
        else:
            return startDate     
```

And the exception class: 

```
class InValidStartDate(Exception):
    """Exception raised for errors in input start date.

    Attributes:
        start_date - input salary which caused the error
        message - explanation of the error
    """
    
    def __init__(self, start, message="Invalid start date entered. Please enter a valid start date."):
        self.start = start
        self.message = message
        super().__init__(self.message)
```
 
Adding tests at the end to check if the exception handling was correct: 

```
# Testing whether exception handling works:
# Pass case:
planPass = Plan('My Plan',
                emergency_type=Plan.EmergencyType.EARTHQUAKE,
                description='Test emergency plan',
                geographical_area='',
                start_date="09/11/2022")

print(planPass.check_date(start_date="09/11/2022"))

# Fail case:
planFail = Plan('My Plan',
                emergency_type=Plan.EmergencyType.EARTHQUAKE,
                description='Test emergency plan',
                geographical_area='',
                start_date="09/11/2020")

print(planFail.check_date(start_date="09/11/2020"))
```

Thank you! Everything else looks perfect.

dennisylyung commented:
> Ensuring that start_date is valid by including the following in class Plan:
>
> ...
> 
> Thank you! Everything else looks perfect.

@Shalaka-Yerawadekar if we are only checking for start dates earlier than today, I think we can be more specific in the exception name and the message. 
Apart from that, do we also need to check for invalid date formats? (i.e. such that `datetime.datetime.strptime` throws an error. ) Alternatively, we can check for string format issues only on the ui level, so that the input to this class is a `datetime.date` instance instead of a date string.

Shalaka-Yerawadekar commented:
@dennisylyung That's a good point, I'll change the exception name and message in that case. 

```
 def check_date(self, start_date):
        startDate = datetime.datetime.strptime(start_date, "%d/%m/%Y")
        present = datetime.datetime.now()
        if startDate < present:
            raise PastStartDate(start_date)
        else:
            return startDate   
            
class  PastStartDate(Exception):
    """Exception raised when start date entered is in the past.

    Attributes:
        start_date - input start_date which caused the error
        message - explanation of the error
    """
    
    def __init__(self, start, message="Start date is in the past. Please enter a valid start date."):
        self.start = start
        self.message = message
        super().__init__(self.message)
```

How about something like the above? 

As for the second point, I'll try handling invalid date formats and get back to you.

Shalaka-Yerawadekar commented on models/plan.py:
```diff
@@ -0,0 +1,36 @@
+        self.geographical_area = geographical_area
+        self.start_date = start_date
+        self.camps: set[Camp] = set(camps) if camps else set()
+
```
Changes based on feedback: 

Ensuring that start_date is valid by including the following in class Plan:
```
    def check_date(self, start_date):
        try:
            startDate = datetime.datetime.strptime(start_date, "%d/%m/%Y")
            present = datetime.datetime.now()
            if startDate < present:
                raise PastStartDate(start_date)
            else:
                return startDate
        except:
            raise ValidStartDateString(start_date)
```

And the exception classes: 

```

# Exception handling for start date:
class PastStartDate(Exception):
    """Exception raised when start date entered is in the past.

    Attributes:
        start_date - input start_date which caused the error
        message - explanation of the error
    """

    def __init__(self, start, message="Start date is in the past. Please enter a valid start date."):
        self.start = start
        self.message = message
        super().__init__(self.message)


class ValidStartDateString(Exception):
    """Exception raised when start date entered is not a valid string input.

    Attributes:
        start_date - input start_date which caused the error
        message - explanation of the error
    """

    def __init__(self, start, message="Invalid start date entered. Please enter a valid string of the form: dd/mm/yy."):
        self.start = start
        self.message = message
        super().__init__(self.message)

```

Adding tests at the end to check if the exception handling was correct:

```
# Testing whether exception handling works:
# Pass case:
planPass = Plan('My Plan',
                emergency_type=Plan.EmergencyType.EARTHQUAKE,
                description='Test emergency plan',
                geographical_area='',
                start_date="09/11/2022")

print(planPass.check_date(start_date="09/11/2022"))

# PastStartDate error:
planFail = Plan('My Plan',
                emergency_type=Plan.EmergencyType.EARTHQUAKE,
                description='Test emergency plan',
                geographical_area='',
                start_date="09/11/2020")

print(planFail.check_date(start_date="09/11/2020"))

#ValidStartDateString error:
planFail2 = Plan('My Plan',
                emergency_type=Plan.EmergencyType.EARTHQUAKE,
                description='Test emergency plan',
                geographical_area='',
                start_date="random string")

```

dennisylyung commented:
@Shalaka-Yerawadekar I tried to merge the date validation part into my code, with some modifications:
1. Simplifying the exception class: For now, I think the attributes are not used.
2. Removed date string exception catching: I think we may better defer the handling to the interface
3. Make unit tests: I changed your test codes into formal unittest methods

What do you think?

Shalaka-Yerawadekar commented:
@dennisylyung Thanks for merging the changes together and creating a formal unit test. I'm happy with all of the changes.

dennisylyung commented:
@Elynor-Kamil just pushed the changes to enforce non empty camps at initialisation, please check~
(P.S. you can view changes in a single commit in the "Files Changed" pane)

michellechan1123 reviewed (commented):
For the plan start date, I am not sure if we can forecast plans in advance, since one possibility will be what if the plan is being created but not used?

```
def __check_start_date(self, start_date: date) -> None:
        """
        Validate start date. PastStartDateException is raised if start date is in the past.
        """
        today = datetime.today().date()
        if start_date < today:
            raise self.PastStartDateException()
```

vwcchan commented on models/plan.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        Close one or more camps.
+        """
+        self.camps.difference_update(camps)
```
I understand that using `difference_update()` can be more efficient than using `remove()` and `discard()` with iteration, but if the admin enter a set of camps to be closed and some of the camps entered do not exist, this function wouldn't show the admin about this (i.e. the admin wouldn't know he entered an invalid camp name). I am wondering if we should add a message for this in this function?

dennisylyung replied:
Good point, I agree that it would make more sense to raise an error when someone tries to close a camp that is not in the plan. 
I have also put that in the method docstring explicitly

dennisylyung commented:
> For the plan start date, I am not sure if we can forecast plans in advance, since one possibility will be what if the plan is being created but not used?
> 
> ```
> def __check_start_date(self, start_date: date) -> None:
>         """
>         Validate start date. PastStartDateException is raised if start date is in the past.
>         """
>         today = datetime.today().date()
>         if start_date < today:
>             raise self.PastStartDateException()
> ```

I agree that scheduling to start a plan may not be too useful, but add complications to the system. 
I checked that this is not a requirement in the coursework brief. 
I suggest setting the start date automatically at creation as a record, and also making it read-only by making it private. 
@Elynor-Kamil @Shalaka-Yerawadekar What do you think?

Shalaka-Yerawadekar commented:
That's actually a very good point, thank you for spotting that @michellechan1123! I'm happy with changing it as you suggested @dennisylyung.

dennisylyung commented:
@michellechan1123 updated to make start date fixed to plan creation date and read-only
